proposals/0000-markdown-content-redesign.md

```diff
@@ -0,0 +1,126 @@
+- Start Date: 03-10-2022
+- Reference Issues: https://github.com/withastro/rfcs/discussions/5, https://github.com/withastro/rfcs/discussions/118
+- Implementation PR: <!-- leave empty -->
+
+# Summary
+
+A v2.0 API for working with local Markdown files in Astro.
+
+# Example
+
+```astro
+---
+// BASIC USAGE:
+
+// Note: Astro.fetchContent() removed in favor of direct `import.meta.globEager` usage
+// Note: This works in non-Astro framework components!
+const markdownFilesObj = import.meta.globEager('../posts/*.md');
+const markdownFilesArr = Object.values(markdownFilesObj);
+const markdownFilesFiltered = markdownFilesArr.filter(post => post.data.category === 'blog');
+
+// Note: Individual markdown imports are now supported as well.
+const firstPost = await import('../posts/first-post.md');
+---
+<!-- Data is parsed from frontmatter without requiring a render -->
+<h1>{firstPost.data.title}</h1>
+<p>Author: {firstPost.data.author}</p>
+<p><a href={firstPost.url}>Permalink</a></p>
+<!-- Defer rendering as late as possible for performance -->
+<article>{firstPost.getContent()}</article>
+```
+
+
+# Motivation
+
+- **Performance:** Markdown rendering continues to be an expensive part of Astro's runtime. While we can look into faster renderers (ex: Goldmark) there is still a big inefficincy on our end with how we hold the tool. Astro currently renders markdown on import, which means that multi-file markdown imports (via both `import.meta.glob` and `Astro.fetchContent()`) block the response until all imported files are rendered, even if only a subset of these files are ever used on the page after filtering.
+- **Performance (Memory):** For large projects, this also forces Astro to store all rendered Markdown output in memory at the same time, making it difficult to manage memory efficiently. Our community has reported builds maxing out memory limits in a project of only 800 markdown pages, and builds requiring 50+ GB of memory.
+- **Infinite Loops:** By rendering during import, we also introduce a risk for infinite loops when we call `fetchContent()` in a Markdown layout used by a fetched page, bouncing between render and `fetchContent()` infinitely.
+
+
+# Detailed design
+
+- This is implemented in a v2.0 of the internal `'astro:markdown'` Vite plugin
+- All Markdown files (ex: `src/posts/foo.md`) are resolved and loaded by this plugin.
+
+1. `src/posts/foo.md?import`
+    1. loaded from file system
+    2. frontmatter is parsed from the file
+    3. a JS module is returned with the following JS:
+    ```js
+    `
+    // Static:
+    export const data = ${JSON.stringify(frontmatter)};
```

Comment:
🚲 since this is specific to markdown files, `frontmatter` is a little more clear for me than `data`

🚲 === we need a bike shedding emoji

Reply:
👍 In updating the examples, I realized that `.data` is used in our pagination API. Not a huge deal, but based on your feedback + my experience migrating that file and adding a ton of `.data` everywhere, I'm +1 on `frontmatter` now as well.